src/Strategy/AlpineLinux/Parser.hs

```diff
@@ -0,0 +1,128 @@
+import Data.List.NonEmpty qualified as NE
+import Data.Map qualified as Map
+import Data.String.Conversion (toText)
+import Data.Text (Text)
+import Data.Void (Void)
+import Strategy.AlpineLinux.Types (AlpinePackage (..))
+import Text.Megaparsec (
+  MonadParsec (eof, takeWhileP),
+  Parsec,
+  many,
+  optional,
+  some,
+  try,
+ )
+import Text.Megaparsec.Char (char, eol, letterChar)
+
+-- | Errors that can occur even in a valid file.  Packages are a lose collection
+-- of fields, but it unclear which are required.  These error represent fields
+-- we expect, but which may be missing in unusual circumstances.
+data PackageError
+  = MissingPackageName
+  | MissingPackageArchitecture {packageName :: Text}
+  | MissingPackageVersion {packageName :: Text}
+  deriving (Show, Eq, Ord)
+
+type Parser = Parsec Void Text
+
+-- | Key for the package-name field.
+packageNameKey :: Text
+packageNameKey = "P"
+
+-- | Key for the architecture field.
+architectureKey :: Text
+architectureKey = "A"
+
+-- | Key for the version field.
+versionKey :: Text
+versionKey = "V"
+
+-- | Parses zero or more occurrences of a construct with separators between.
+--
+-- This differs from `Control.Monad.Combinators.sepBy` because it will backtrack
+-- failed repetitions.  The generic version has no notion of backtracking so if
+-- the separator consumes input after the last entry then the whole parser will
+-- fail because it will fail to parse another entry after the separator.
+--
+-- TODO: If this is more generally useful we should move it to a common location.
+--
+-- The Megaparsec project chose not to include this in their library:
+--    https://github.com/mrkkrp/megaparsec/issues/401
+-- This implementation is written to mirror the source of the generic version:
+--   https://hackage.haskell.org/package/parser-combinators-1.3.0/docs/src/Control.Monad.Combinators.html#sepBy
+sepByWithTry :: MonadParsec e s m => m a -> m b -> m [a]
+sepByWithTry p sep = do
+  r <- optional (try p)
+  case r of
+    Nothing -> pure []
+    Just x -> (x :) <$> many (try (sep >> p))
+
+-- | Parses one or more occurrences of a construct with separators between.
+--
+-- See `sepByWithTry` for an explanation and motivation.
+--
+-- This implementation is written to mirror the source of the generic version:
+--   https://hackage.haskell.org/package/parser-combinators-1.3.0/docs/src/Control.Monad.Combinators.html#sepBy1
+sepByWithTry1 :: MonadParsec e s m => m a -> m b -> m (NonEmpty a)
+sepByWithTry1 p sep = do
+  x <- try p
+  (x :|) <$> many (try (sep >> p))
+
+-- | Parses an Alpine Linux installed-packages database file.
+-- This returns a list of packages or errors that represent malformed packages.
+--
+-- Packages are separated from one another by one-or-more newlines.  The end of
+-- a file may have any number of trailing newlines.
+installedPackagesDatabaseParser :: Parser [Either PackageError AlpinePackage]
+installedPackagesDatabaseParser =
```

Comment:
[Optional, Nit] might be a good idea to have a lexer (or trim them after), right now adding a space after the value in the field will break the parser. Folks may have modified this file manually, or other tools may have made modifications - good to stay on side of caution. 

```haskell
-- note the space after version value
P:package-name\r\nA:x86-64\r\nV:some-version \r\n\r\n
```

Reply:
I think separating the lexer and parser stages is a great way to solve the oddities of this parser code that arise from the oddities of the format.

Reply:
The try-based `sepBy` combinator might not be needed if operating on a `Seq ApkPackageToken`.

Reply:
Trimming makes sense.  That's pretty easy to put in.  I'm not sure we need a full-blown lexer though.

Reply:
The phrase "full-blown" makes it sound more difficult than I think it is (or less valuable).  I think that separating the two makes each part simpler, and most of the work you've done handles the lexing part anyway.

I'd like to see more separation of lexing and parsing in the codebase in general, specifically because it makes each part of the code much simpler to understand and is primarily useful in desugaring these overloaded tokens: a lexer could make a clear differentiation between a single new line and 2+ newlines, by tokenizing them as `PropertySeparator` and `PackageSeparator`.

It also allows testing individual parts of the parsing scheme easier.  Testing the details of this is a little clearer than E2E testing a lexing parser:
```haskell
type Lexer = Parsec Void Text
type Parser = Parsec PackageError [ApkToken]

lexer :: Lexer [ApkToken]
...

parser :: Parser AlpinePackage
...

doTheWholeThing someInput = do
  -- Won't actually work out-of-box since the Eithers have different Left types
  apkTokens <- Megaparsec.parse lexer someInput
  package <- Megaparsec.parse parser apkTokens
```

I don't think this is necessarily simpler overall, but its parts are simpler, and we should probably be doing this with our parsers in general.